File: src/lib/grid/allocation.ts
/**
 * GridPulse charging allocation — a deterministic optimization/control
 * algorithm. No AI, no learning, no black box: identical inputs always
 * produce identical decisions, and every decision carries the factors and
 * reasons that produced it so the UI can explain itself.
 *
 * ─── How it works ──────────────────────────────────────────────────────────
 *
 * 1. BUDGET. Available charging capacity for the EV cluster:
 *
 *      availableChargingCapacity = gridCapacity - buildingLoad + renewable
 *
 *    Building demand is always served first and is never interrupted. On-site
 *    solar displaces grid import kW-for-kW and the transformer carries only
 *    the import, so this budget can never violate the grid limit — see the
 *    invariant derivation in Step 1 below. Surplus solar legitimately adds
 *    charging headroom instead of being wasted. The result is clamped at 0;
 *    the allocator never invents power.
 *
 * 2. PRIORITY. Each vehicle gets a 0-100 composite score from three
 *    independently normalized factors (see scoreVehicle / priorityFactors):
 *      - batteryUrgency    how low the battery is right now
 *      - departureUrgency  how close departure is, including whether there
 *                          is still enough TIME to finish charging at all
 *      - energyDeficit     how much charge is still missing
 *
 * 3. DISTRIBUTION. Two phases:
 *      Phase A — priority service: vehicles are served at FULL rate in
 *                priority order while headroom allows.
 *      Phase B — weighted fair share: once one vehicle can no longer be
 *                fully served, the remaining headroom is split among it and
 *                all lower-priority vehicles proportionally to their
 *                priority scores (water-filling, capped at each vehicle's
 *                max, re-iterating to redistribute surplus). Scarcity is
 *                shared instead of first-come-first-served.
 *
 * 4. EXPLAIN. Every allocation records its normalized factors, serving
 *    order, the budget it was decided under, its status class, and honest
 *    reason strings naming the constraint that actually bound the decision.
 */
import type { Allocation, AllocationResult, ChargingStatus, GridInputs, Vehicle } from "./types";

/** Earliest sensible departure used when a vehicle time is missing/malformed. */
const DEFAULT_DEPARTURE_MINUTES = 12 * 60;

/**
 * Departure urgency saturates this many minutes out. 8h chosen so a vehicle
 * leaving in the morning is already near-max urgent late at night, but a
 * vehicle leaving next shift does not dominate the queue unfairly.
 */
const DEPARTURE_HORIZON_MINUTES = 480;

/**
 * Below this state of charge a vehicle is critically low and receives
 * maximum battery urgency regardless of other factors.
 */
const CRITICAL_BATTERY = 20;

/**
 * Departure urgency saturates once there is no longer enough time to charge
 * to target before departure — a real deadline miss looming.
 */
const DEADLINE_RISK_BUFFER_MINUTES = 15;

export function parseDeparture(departure: string): number {
  const parts = departure.split(":").map(Number);
  const hours = Number.isFinite(parts[0]) ? (parts[0] as number) : NaN;
  const minutes = Number.isFinite(parts[1]) ? (parts[1] as number) : NaN;
  if (
    !Number.isFinite(hours) ||
    !Number.isFinite(minutes) ||
    hours < 0 ||
    hours > 23 ||
    minutes < 0 ||
    minutes > 59
  ) {
    return DEFAULT_DEPARTURE_MINUTES;
  }
  return hours * 60 + minutes;
}

/** Minutes since midnight on the 24h clock. */
export function minutesOfDay(date: Date): number {
  return date.getHours() * 60 + date.getMinutes();
}

/** Clamp to [min, max]. */
function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}

/** Round to one decimal for display-stable numbers. */
function round1(value: number): number {
  return Number(value.toFixed(1));
}

/**
 * Floor to one decimal. Used for granted power so that rounded grants can
 * never sum ABOVE the true budget — the grid invariant survives rounding.
 */
function floor1(value: number): number {
  return Math.max(0, Math.floor(value * 10) / 10);
}

/**
 * kWh still required for a vehicle to reach its target state of charge:
 * (target% - battery%) x pack kWh / 100.
 */
function energyNeededKwh(vehicle: Vehicle): number {
  const deficitPercent = Math.max(0, vehicle.target - vehicle.battery);
  return (deficitPercent / 100) * vehicle.batteryPackKwh;
}

/**
 * Minutes of charging at `powerKw` needed to reach target. Null when power
 * is zero or no energy is needed (no estimate is possible/required).
 */
function minutesToTarget(energyKwh: number, powerKw: number): number | null {
  if (powerKw <= 0 || energyKwh <= 0) return null;
  return (energyKwh / powerKw) * 60;
}

/**
 * PRIORITY SCORING.
 *
 * Three factors, each independently normalized to 0-1 with explicit,
 * documented anchors — the weights are the only policy inputs:
 *
 *   batteryUrgency    (100 - battery) / (100 - CRITICAL_BATTERY), capped at
 *                     1. A near-empty battery is the strongest single
 *                     signal; at or below 20% it saturates.
 *
 *   departureUrgency  1 - clamp(minutesToDeparture / HORIZON, 0, 1). Linear
 *                     ramp over an 8h horizon. When charging to target can
 *                     no longer FINISH before departure (plus a small
 *                     buffer) even at max power, urgency is forced to 1 —
 *                     a deadline is a deadline no matter what the ramp says.
 *
 *   energyDeficit     (target - battery) / target. Relative deficit, so a
 *                     40%->80% gap outweighs an 85%->90% top-up.
 *
 * Composite score = 40% battery + 35% departure + 25% deficit, as an
 * integer 0-100. Weights are policy, stated openly here.
 */
export function scoreVehicle(vehicle: Vehicle, nowMinutes: number): number {
  const { batteryUrgency, departureUrgency, energyDeficit } = priorityFactors(vehicle, nowMinutes);
  return Math.round((batteryUrgency * 0.4 + departureUrgency * 0.35 + energyDeficit * 0.25) * 100);
}

/** Compute the raw normalized (0-1) priority factors for one vehicle. */
function priorityFactors(vehicle: Vehicle, nowMinutes: number) {
  const departureMinutes = parseDeparture(vehicle.departure);

  // Time until departure; a vehicle past its departure time is maximally
  // urgent (zero minutes left).
  const minutesToDeparture = Math.max(0, departureMinutes - nowMinutes);

  const batteryUrgency = clamp((100 - vehicle.battery) / (100 - CRITICAL_BATTERY), 0, 1);

  let departureUrgency = 1 - clamp(minutesToDeparture / DEPARTURE_HORIZON_MINUTES, 0, 1);

  // Deadline risk: if charging at max power cannot reach target before
  // departure (with a small buffer), urgency saturates at 1.
  const needed = energyNeededKwh(vehicle);
  const fastestMinutes = minutesToTarget(needed, vehicle.maxPower);
  if (
    fastestMinutes !== null &&
    minutesToDeparture > 0 &&
    fastestMinutes + DEADLINE_RISK_BUFFER_MINUTES >= minutesToDeparture
  ) {
    departureUrgency = 1;
  }
  // Already past departure with energy still missing: maximally urgent.
  if (minutesToDeparture === 0 && needed > 0) departureUrgency = 1;

  const energyDeficit =
    vehicle.target > 0 ? clamp((vehicle.target - vehicle.battery) / vehicle.target, 0, 1) : 0;

  return { batteryUrgency, departureUrgency, energyDeficit };
}

/**
 * Priority explanation in plain language. Reads the same normalized factors
 * the score used and names the dominant one — the operator should never
 * have to reverse-engineer a number.
 */
function priorityReason(vehicle: Vehicle, factors: ReturnType<typeof priorityFactors>): string {
  const { batteryUrgency, departureUrgency, energyDeficit } = factors;
  if (vehicle.battery <= CRITICAL_BATTERY) {
    return `Critical battery (${vehicle.battery}%) — near-empty vehicles are served first.`;
  }
  if (departureUrgency >= 0.95) {
    return `Departure at ${vehicle.departure} is close and there may not be enough time to finish charging.`;
  }
  if (departureUrgency >= 0.6) {
    return `Departure at ${vehicle.departure} is approaching — needs its charge soon.`;
  }
  if (batteryUrgency >= 0.6) {
    return `Battery is getting low (${vehicle.battery}%).`;
  }
  if (energyDeficit >= 0.5) {
    return `Large remaining charge (${vehicle.battery}% → ${vehicle.target}%).`;
  }
  return `Comfortable margins — battery ${vehicle.battery}%, departs ${vehicle.departure}.`;
}

/**
 * THE ALLOCATOR.
 *
 * Deterministic: same vehicles + inputs + clock produce the same output.
 * Sorting is score-descending with id tiebreak, so ties never flip between
 * renders. Grid safety invariant: building load is served first and total
 * EV draw never exceeds the available charging capacity.
 */
export function allocate(
  vehicles: Vehicle[],
  inputs: GridInputs,
  nowMinutes: number,
): AllocationResult {
  const { gridCapacity, buildingDemand, solarGeneration } = inputs;

  // ── Step 1: cluster budget ────────────────────────────────────────────────
  // availableChargingCapacity = gridCapacity - buildingLoad + renewable.
  // On-site solar displaces grid import kW-for-kW, and the transformer only
  // carries the IMPORT, so:
  //   import = building + ev - solar  <=  gridCapacity
  //   <=>    ev        <=  gridCapacity - building + solar = budget
  // The invariant is therefore satisfied by construction — the EV cluster can
  // never overload the transformer, even during a solar surplus. The result
  // is clamped at 0; the allocator never invents power.
  const availableChargingCapacityKw = Math.max(0, gridCapacity - buildingDemand + solarGeneration);
  const clusterHasCapacity = availableChargingCapacityKw > 1e-9;

  // ── Step 2: deterministic priority order ─────────────────────────────────
  const scored = vehicles
    .map((vehicle) => ({
      vehicle,
      factors: priorityFactors(vehicle, nowMinutes),
      score: scoreVehicle(vehicle, nowMinutes),
    }))
    .sort(
      (a, b) =>
        b.score - a.score ||
        (a.vehicle.id < b.vehicle.id ? -1 : a.vehicle.id > b.vehicle.id ? 1 : 0),
    );

  const needed = (entry: (typeof scored)[number]) => energyNeededKwh(entry.vehicle);

  // Shared builder so every allocation carries the full explanation payload.
  const buildAllocation = (
    entry: (typeof scored)[number],
    grantedPrecise: number,
    status: ChargingStatus,
    phaseBBudgetKw: number,
    peerCount: number,
  ): Allocation => {
    const { vehicle, factors, score } = entry;
    const energyNeeded = needed(entry);
    // Floor to 0.1 kW: rounding must never push total draw above the budget.
    const power = floor1(grantedPrecise);
    const etaMinutes = minutesToTarget(energyNeeded, power);
    const minutesToDeparture = Math.max(0, parseDeparture(vehicle.departure) - nowMinutes);
    const notes: string[] = [];
    if (
      power > 0 &&
      etaMinutes !== null &&
      minutesToDeparture > 0 &&
      etaMinutes + DEADLINE_RISK_BUFFER_MINUTES >= minutesToDeparture
    ) {
      notes.push(
        `At ${power} kW it will not reach ${vehicle.target}% before its ${vehicle.departure} departure.`,
      );
    }
    return {
      ...vehicle,
      priority: score,
      priorityFactors: factors,
      power,
      energyNeededKwh: round1(energyNeeded),
      status,
      etaMinutesFromNow: etaMinutes === null ? null : Math.max(1, Math.round(etaMinutes)),
      capacityBudgetKw: round1(availableChargingCapacityKw),
      servedOrder: -1, // assigned after final assembly
      priorityReason: priorityReason(vehicle, factors),
      powerReason: powerReason({
        granted: power,
        vehicle,
        status,
        phaseBBudgetKw,
        peerCount,
      }),
      notes,
    };
  };

  // Completed vehicles need nothing — they simply free their share.
  const completedAllocs: Allocation[] = scored
    .filter((entry) => needed(entry) <= 1e-9)
    .map((entry) => ({
      ...entry.vehicle,
      priority: entry.score,
      priorityFactors: entry.factors,
      power: 0,
      energyNeededKwh: 0,
      status: "completed" as const,
      etaMinutesFromNow: null,
      capacityBudgetKw: round1(availableChargingCapacityKw),
      servedOrder: -1,
      priorityReason: `Target reached — battery ${entry.vehicle.battery}% ≥ target ${entry.vehicle.target}%.`,
      powerReason: "Charging complete — no energy required.",
      notes: [],
    }));

  const active = scored.filter((entry) => needed(entry) > 1e-9);
  const served: Allocation[] = [];
  const unserved: Allocation[] = [];

  let remaining = availableChargingCapacityKw;

  // ── Step 3a: priority service (Phase A) ──────────────────────────────────
  // Serve vehicles at FULL rate in priority order while headroom allows.
  // Stops at the first vehicle the remaining headroom cannot fully serve.
  let cursor = 0;
  while (cursor < active.length) {
    const entry = active[cursor];
    if (!entry || remaining < entry.vehicle.maxPower - 1e-9) break;
    served.push(buildAllocation(entry, entry.vehicle.maxPower, "charging", 0, 0));
    remaining = Math.max(0, remaining - entry.vehicle.maxPower);
    cursor++;
  }

  // ── Step 3b: weighted fair share (Phase B) ───────────────────────────────
  // The first unservable vehicle and everyone below it share what is left,
  // proportionally to priority score (weight floor 1 keeps zero-score
  // vehicles in the pool). Water-filling: each round computes desired shares
  // against the round's opening pool, applies them in priority order, and
  // vehicles already at their max release surplus to the next round.
  const phaseBBudgetKw = remaining;
  const pending = active.slice(cursor).map((entry) => ({ entry, granted: 0 }));

  for (let round = 0; round <= pending.length && remaining > 1e-9; round++) {
    const open = pending.filter((p) => p.entry.vehicle.maxPower - p.granted > 1e-9);
    if (!open.length) break;

    const totalWeight = open.reduce((sum, p) => sum + Math.max(1, p.entry.score), 0);
    const snapshot = remaining;
    let pool = snapshot;
    let progressed = false;

    for (const p of open) {
      const weight = Math.max(1, p.entry.score);
      const want = p.entry.vehicle.maxPower - p.granted;
      const share = (snapshot * weight) / totalWeight;
      const give = Math.min(want, share);
      if (give > 1e-9) {
        p.granted += give;
        pool -= give;
        progressed = true;
      }
    }

    remaining = Math.max(0, pool);
    if (!progressed) break;
  }

  for (const p of pending) {
    const status = classifyStatus({
      granted: p.granted,
      vehicle: p.entry.vehicle,
      sharingWithPeers: pending.length > 1,
      clusterHasCapacity,
    });
    const target = p.granted > 1e-9 ? served : unserved;
    target.push(
      buildAllocation(p.entry, p.granted, status, phaseBBudgetKw, Math.max(0, pending.length - 1)),
    );
  }

  // Final order: actively charging first (serving order), then queued
  // vehicles, then completed — each group priority-ordered.
  const allocations = [...served, ...unserved, ...completedAllocs];
  for (let index = 0; index < allocations.length; index++) {
    const allocation = allocations[index];
    if (allocation && allocation.power > 0) allocation.servedOrder = index;
  }

  // ── Step 4: cluster-level facts for the UI ───────────────────────────────
  const totalAllocatedKw = round1(served.reduce((sum, a) => sum + a.power, 0));
  return {
    allocations,
    availableChargingCapacityKw: round1(availableChargingCapacityKw),
    totalAllocatedKw,
    chargingCount: served.length,
    completedCount: completedAllocs.length,
  };
}

/**
 * Status classification. Distinguishes the flavors of "getting power":
 * full rate (charging), reduced because capacity is shared with peer
 * vehicles (throttled), and reduced because the grid itself cannot offer
 * more even to a single vehicle (constraint-limited). And the flavors of
 * "not getting power": queued while capacity exists (scheduled), zero
 * capacity anywhere (waiting), and finished (completed).
 */
function classifyStatus(args: {
  granted: number;
  vehicle: Vehicle;
  sharingWithPeers: boolean;
  clusterHasCapacity: boolean;
}): ChargingStatus {
  const { granted, vehicle, sharingWithPeers, clusterHasCapacity } = args;
  if (granted >= vehicle.maxPower - 1e-9) return "charging";
  if (granted > 1e-9) return sharingWithPeers ? "throttled" : "constraint-limited";
  return clusterHasCapacity ? "scheduled" : "waiting";
}

/**
 * Plain-language explanation of WHY the vehicle got its power level. Honest
 * and specific — it names the constraint that actually bound the decision.
 */
function powerReason(args: {
  granted: number;
  vehicle: Vehicle;
  status: ChargingStatus;
  phaseBBudgetKw: number;
  peerCount: number;
}): string {
  const { granted, vehicle, status, phaseBBudgetKw, peerCount } = args;
  switch (status) {
    case "charging":
      return `Full ${vehicle.maxPower} kW rate — grid headroom allowed it.`;
    case "throttled":
      return `Charging reduced to ${granted} kW of ${vehicle.maxPower} kW — the remaining ${round1(phaseBBudgetKw)} kW of headroom is shared with ${peerCount} other vehicle${peerCount === 1 ? "" : "s"}, weighted by priority.`;
    case "constraint-limited":
      return `Capped at ${granted} kW — even alone, the grid could only offer ${round1(phaseBBudgetKw)} kW beyond building demand right now.`;
    case "scheduled":
      return `Queued behind higher-priority vehicles — charging begins as they finish or depart.`;
    case "waiting":
      return `No charging capacity available — building demand is using the entire grid budget. Charging starts when headroom returns.`;
    case "completed":
      return `Charging complete — no energy required.`;
  }
}

/**
 * Human-readable charging completion estimate, or a short honest
 * explanation of why no estimate is possible.
 */
export function formatEta(allocation: Allocation): string {
  if (allocation.status === "completed") return "Done";
  if (allocation.etaMinutesFromNow === null) {
    return allocation.power > 0 ? "—" : "Not charging";
  }
  const minutes = allocation.etaMinutesFromNow;
  if (minutes < 60) return `~${minutes} min`;
  const hours = Math.floor(minutes / 60);
  const rest = minutes % 60;
  return rest === 0 ? `~${hours} h` : `~${hours} h ${rest} min`;
}

File: backend/allocation.py
"""GridPulse charging allocation engine.

Deterministic optimization/control logic ported 1:1 from the TypeScript
implementation (src/lib/grid/allocation.ts). Identical inputs always produce
identical decisions; every decision carries the factors and reasons that
produced it so the UI can explain itself.

Grid safety invariant:
    building demand + EV charging - solar generation <= grid capacity

The cluster budget is derived as:
    available = max(0, grid_capacity - building_demand + solar_generation)
so the invariant holds by construction.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Any

# ── Policy constants (identical to the TypeScript engine) ────────────────────
DEPARTURE_HORIZON_MINUTES = 480  # 8h linear ramp for departure urgency
CRITICAL_BATTERY = 20  # battery% at/below which urgency saturates
DEADLINE_RISK_BUFFER_MINUTES = 15  # deadline-risk buffer before departure
DEFAULT_DEPARTURE_MINUTES = 12 * 60  # 12:00 fallback for malformed times

EPS = 1e-9

CHARGING_STATUSES = (
    "charging",
    "waiting",
    "scheduled",
    "completed",
    "throttled",
    "constraint-limited",
)


def clamp(value: float, lo: float, hi: float) -> float:
    return min(hi, max(lo, value))


def round1(value: float) -> float:
    """Round to one decimal for display-stable numbers."""
    return round(value, 1)


def floor1(value: float) -> float:
    """Floor to one decimal so rounded grants never sum ABOVE the budget."""
    return max(0.0, math.floor(value * 10) / 10)


def js_num(value: float) -> str:
    """Format a float the way JavaScript's default Number->string does.

    3.0 -> "3", 3.5 -> "3.5", 14.0 -> "14". Used inside reason strings so
    they render identically to the original TypeScript UI.
    """
    if value == int(value) and abs(value) < 1e21:
        return str(int(value))
    return repr(value)


def parse_departure(departure: str) -> int:
    """Parse 'HH:MM' into minutes since midnight, with a sane fallback."""
    try:
        parts = departure.split(":")
        if len(parts) != 2:
            return DEFAULT_DEPARTURE_MINUTES
        hours, minutes = int(parts[0]), int(parts[1])
    except (ValueError, AttributeError):
        return DEFAULT_DEPARTURE_MINUTES
    if not (0 <= hours <= 23 and 0 <= minutes <= 59):
        return DEFAULT_DEPARTURE_MINUTES
    return hours * 60 + minutes


def minutes_of_day(now_minutes: int | None = None) -> int:
    return now_minutes if now_minutes is not None else 0


def energy_needed_kwh(vehicle: dict[str, Any]) -> float:
    """kWh still required to reach target: (target% - battery%) x pack / 100."""
    deficit_percent = max(0.0, vehicle["target"] - vehicle["battery"])
    return (deficit_percent / 100) * vehicle["batteryPackKwh"]


def minutes_to_target(energy_kwh: float, power_kw: float) -> float | None:
    """Minutes of charging at power_kw needed to reach target.

    None when power is zero or no energy is needed.
    """
    if power_kw <= 0 or energy_kwh <= 0:
        return None
    return (energy_kwh / power_kw) * 60


def priority_factors(vehicle: dict[str, Any], now_minutes: int) -> dict[str, float]:
    """Raw normalized (0-1) priority factors for one vehicle."""
    departure_minutes = parse_departure(vehicle["departure"])

    # Time until departure; a vehicle past its departure time is maximally
    # urgent (zero minutes left).
    minutes_to_departure = max(0, departure_minutes - now_minutes)

    battery_urgency = clamp(
        (100 - vehicle["battery"]) / (100 - CRITICAL_BATTERY), 0, 1
    )

    departure_urgency = 1 - clamp(
        minutes_to_departure / DEPARTURE_HORIZON_MINUTES, 0, 1
    )

    # Deadline risk: if charging at max power cannot reach target before
    # departure (with a small buffer), urgency saturates at 1.
    needed = energy_needed_kwh(vehicle)
    fastest_minutes = minutes_to_target(needed, vehicle["maxPower"])
    if (
        fastest_minutes is not None
        and minutes_to_departure > 0
        and fastest_minutes + DEADLINE_RISK_BUFFER_MINUTES >= minutes_to_departure
    ):
        departure_urgency = 1
    # Already past departure with energy still missing: maximally urgent.
    if minutes_to_departure == 0 and needed > 0:
        departure_urgency = 1

    energy_deficit = (
        clamp((vehicle["target"] - vehicle["battery"]) / vehicle["target"], 0, 1)
        if vehicle["target"] > 0
        else 0.0
    )

    return {
        "batteryUrgency": battery_urgency,
        "departureUrgency": departure_urgency,
        "energyDeficit": energy_deficit,
    }


def score_vehicle(vehicle: dict[str, Any], now_minutes: int) -> int:
    """Composite priority score 0-100.

    40% battery + 35% departure + 25% deficit. Weights are policy, stated
    openly here.
    """
    factors = priority_factors(vehicle, now_minutes)
    composite = (
        factors["batteryUrgency"] * 0.4
        + factors["departureUrgency"] * 0.35
        + factors["energyDeficit"] * 0.25
    )
    # JavaScript Math.round(-0.5) -> 0 (rounds toward +Infinity); Python
    # round() uses banker's rounding. Match the JS behavior exactly.
    return math.floor(composite * 100 + 0.5)


def priority_reason(vehicle: dict[str, Any], factors: dict[str, float]) -> str:
    """Plain-language explanation of WHY the vehicle has its priority."""
    battery_urgency = factors["batteryUrgency"]
    departure_urgency = factors["departureUrgency"]
    energy_deficit = factors["energyDeficit"]

    if vehicle["battery"] <= CRITICAL_BATTERY:
        return (
            f"Critical battery ({vehicle['battery']}%) — near-empty vehicles "
            "are served first."
        )
    if departure_urgency >= 0.95:
        return (
            f"Departure at {vehicle['departure']} is close and there may not "
            "be enough time to finish charging."
        )
    if departure_urgency >= 0.6:
        return (
            f"Departure at {vehicle['departure']} is approaching — needs its "
            "charge soon."
        )
    if battery_urgency >= 0.6:
        return f"Battery is getting low ({vehicle['battery']}%)."
    if energy_deficit >= 0.5:
        return (
            f"Large remaining charge ({vehicle['battery']}% → "
            f"{vehicle['target']}%)."
        )
    return (
        f"Comfortable margins — battery {vehicle['battery']}%, "
        f"departs {vehicle['departure']}."
    )


def classify_status(
    granted: float,
    vehicle: dict[str, Any],
    sharing_with_peers: bool,
    cluster_has_capacity: bool,
) -> str:
    """Status classification, identical semantics to the TypeScript engine."""
    if granted >= vehicle["maxPower"] - EPS:
        return "charging"
    if granted > EPS:
        return "throttled" if sharing_with_peers else "constraint-limited"
    return "scheduled" if cluster_has_capacity else "waiting"


def power_reason(
    granted: float,
    vehicle: dict[str, Any],
    status: str,
    phase_b_budget_kw: float,
    peer_count: int,
) -> str:
    """Plain-language explanation of WHY the vehicle got its power level."""
    if status == "charging":
        return f"Full {vehicle['maxPower']} kW rate — grid headroom allowed it."
    if status == "throttled":
        unit = "vehicle" if peer_count == 1 else "vehicles"
        return (
            f"Charging reduced to {js_num(granted)} kW of {vehicle['maxPower']} kW — "
            f"the remaining {js_num(round1(phase_b_budget_kw))} kW of headroom is "
            f"shared with {peer_count} other {unit}, weighted by priority."
        )
    if status == "constraint-limited":
        return (
            f"Capped at {js_num(granted)} kW — even alone, the grid could only offer "
            f"{js_num(round1(phase_b_budget_kw))} kW beyond building demand right now."
        )
    if status == "scheduled":
        return (
            "Queued behind higher-priority vehicles — charging begins as "
            "they finish or depart."
        )
    if status == "waiting":
        return (
            "No charging capacity available — building demand is using the "
            "entire grid budget. Charging starts when headroom returns."
        )
    return "Charging complete — no energy required."


@dataclass
class _Pending:
    vehicle: dict[str, Any]
    factors: dict[str, float]
    score: int
    granted: float = 0.0


def allocate(
    vehicles: list[dict[str, Any]],
    grid_capacity: float,
    building_demand: float,
    solar_generation: float,
    now_minutes: int,
) -> dict[str, Any]:
    """THE ALLOCATOR — deterministic two-phase allocation.

    Phase A: priority service at full rate while headroom allows.
    Phase B: priority-weighted fair sharing (water-filling) of what remains.
    """
    # ── Step 1: cluster budget ────────────────────────────────────────────
    # available = gridCapacity - buildingLoad + renewable (see module docstring
    # for the invariant derivation). Clamped at 0: never invent power.
    available = max(0.0, grid_capacity - building_demand + solar_generation)
    cluster_has_capacity = available > EPS

    # ── Step 2: deterministic priority order ──────────────────────────────
    scored = sorted(
        (
            {
                "vehicle": vehicle,
                "factors": priority_factors(vehicle, now_minutes),
                "score": score_vehicle(vehicle, now_minutes),
            }
            for vehicle in vehicles
        ),
        key=lambda entry: (-entry["score"], entry["vehicle"]["id"]),
    )

    def build_allocation(
        entry: dict[str, Any],
        granted_precise: float,
        status: str,
        phase_b_budget_kw: float,
        peer_count: int,
    ) -> dict[str, Any]:
        vehicle = entry["vehicle"]
        factors = entry["factors"]
        energy_needed = energy_needed_kwh(vehicle)
        # Floor to 0.1 kW: rounding must never push total draw above budget.
        power = floor1(granted_precise)
        eta_minutes = minutes_to_target(energy_needed, power)
        minutes_to_departure = max(0, parse_departure(vehicle["departure"]) - now_minutes)

        notes: list[str] = []
        if (
            power > 0
            and eta_minutes is not None
            and minutes_to_departure > 0
            and eta_minutes + DEADLINE_RISK_BUFFER_MINUTES >= minutes_to_departure
        ):
            notes.append(
                f"At {js_num(power)} kW it will not reach {vehicle['target']}% before "
                f"its {vehicle['departure']} departure."
            )

        allocation: dict[str, Any] = dict(vehicle)
        allocation.update(
            priority=entry["score"],
            priorityFactors=factors,
            power=power,
            energyNeededKwh=round1(energy_needed),
            status=status,
            etaMinutesFromNow=(
                None if eta_minutes is None else max(1, math.floor(eta_minutes + 0.5))
            ),
            capacityBudgetKw=round1(available),
            servedOrder=-1,
            priorityReason=priority_reason(vehicle, factors),
            powerReason=power_reason(
                power, vehicle, status, phase_b_budget_kw, peer_count
            ),
            notes=notes,
        )
        return allocation

    # ── Completed vehicles need nothing — they free their share ───────────
    completed_allocs = []
    for entry in scored:
        if energy_needed_kwh(entry["vehicle"]) <= EPS:
            vehicle = entry["vehicle"]
            completed_allocs.append(
                {
                    **vehicle,
                    "priority": entry["score"],
                    "priorityFactors": entry["factors"],
                    "power": 0,
                    "energyNeededKwh": 0,
                    "status": "completed",
                    "etaMinutesFromNow": None,
                    "capacityBudgetKw": round1(available),
                    "servedOrder": -1,
                    "priorityReason": (
                        f"Target reached — battery {vehicle['battery']}% ≥ "
                        f"target {vehicle['target']}%."
                    ),
                    "powerReason": "Charging complete — no energy required.",
                    "notes": [],
                }
            )

    active = [entry for entry in scored if energy_needed_kwh(entry["vehicle"]) > EPS]
    served: list[dict[str, Any]] = []
    unserved: list[dict[str, Any]] = []

    remaining = available

    # ── Step 3a: priority service (Phase A) ───────────────────────────────
    cursor = 0
    while cursor < len(active):
        entry = active[cursor]
        if entry is None or remaining < entry["vehicle"]["maxPower"] - EPS:
            break
        served.append(
            build_allocation(entry, entry["vehicle"]["maxPower"], "charging", 0, 0)
        )
        remaining = max(0.0, remaining - entry["vehicle"]["maxPower"])
        cursor += 1

    # ── Step 3b: weighted fair share (Phase B) ────────────────────────────
    phase_b_budget_kw = remaining
    pending: list[_Pending] = [
        _Pending(entry["vehicle"], entry["factors"], entry["score"])
        for entry in active[cursor:]
    ]

    round_index = 0
    while round_index <= len(pending) and remaining > EPS:
        open_pending = [p for p in pending if p.vehicle["maxPower"] - p.granted > EPS]
        if not open_pending:
            break

        total_weight = sum(max(1, p.score) for p in open_pending)
        snapshot = remaining
        pool = snapshot
        progressed = False

        for p in open_pending:
            weight = max(1, p.score)
            want = p.vehicle["maxPower"] - p.granted
            share = (snapshot * weight) / total_weight
            give = min(want, share)
            if give > EPS:
                p.granted += give
                pool -= give
                progressed = True

        remaining = max(0.0, pool)
        if not progressed:
            break
        round_index += 1

    for p in pending:
        status = classify_status(
            granted=p.granted,
            vehicle=p.vehicle,
            sharing_with_peers=len(pending) > 1,
            cluster_has_capacity=cluster_has_capacity,
        )
        entry = {"vehicle": p.vehicle, "factors": p.factors, "score": p.score}
        target = served if p.granted > EPS else unserved
        target.append(
            build_allocation(
                entry,
                p.granted,
                status,
                phase_b_budget_kw,
                max(0, len(pending) - 1),
            )
        )

    # Final order: charging first (serving order), then queued, then done.
    allocations = [*served, *unserved, *completed_allocs]
    for index, allocation in enumerate(allocations):
        if allocation["power"] > 0:
            allocation["servedOrder"] = index

    # ── Step 4: cluster-level facts for the UI ────────────────────────────
    total_allocated = round1(sum(a["power"] for a in served))
    return {
        "allocations": allocations,
        "availableChargingCapacityKw": round1(available),
        "totalAllocatedKw": total_allocated,
        "chargingCount": len(served),
        "completedCount": len(completed_allocs),
    }


def compute_derived(
    grid_capacity: float,
    building_demand: float,
    solar_generation: float,
    ev_load: float,
) -> dict[str, float]:
    """Cluster-level derived metrics (port of derived.ts).

    Import semantics: the transformer carries building + ev - solar. Solar
    can push import below zero (export), clamped where a demand reading
    would be meaningless.
    """
    grid_import = building_demand + ev_load - solar_generation
    total_demand = max(0.0, grid_import)
    headroom = max(0.0, grid_capacity - grid_import)
    utilization = (
        math.floor(max(0.0, grid_import) / grid_capacity * 100 + 0.5)
        if grid_capacity
        else 0
    )
    renewable_share = min(
        100,
        math.floor(solar_generation / max(1, building_demand + ev_load) * 100 + 0.5),
    )
    solar_surplus = max(0.0, solar_generation - building_demand)
    clean_charging = min(ev_load, solar_generation)

    return {
        "gridImport": grid_import,
        "totalDemand": total_demand,
        "headroom": headroom,
        "utilization": utilization,
        "renewableShare": renewable_share,
        "solarSurplus": solar_surplus,
        "cleanCharging": clean_charging,
    }


def format_eta(allocation: dict[str, Any]) -> str:
    """Human-readable completion estimate, or an honest explanation."""
    if allocation["status"] == "completed":
        return "Done"
    eta = allocation.get("etaMinutesFromNow")
    if eta is None:
        return "—" if allocation["power"] > 0 else "Not charging"
    minutes = eta
    if minutes < 60:
        return f"~{minutes} min"
    hours = minutes // 60
    rest = minutes % 60
    return f"~{hours} h" if rest == 0 else f"~{hours} h {rest} min"
